Osd replacement design doc (#17492)
rook/rook: design/ceph/osd-replacement.md

    deactivate Dep
```

### Step-by-step

The walk-through uses the running example.

#### 1. Trigger and validation

The user sets `osd.rook.io/replace=yes-really-replace-osd-<id>` on the OSD Deployment. The numeric `<id>` in the value must equal the Deployment's `ceph-osd-id` label, a copy-paste guard against operating on the wrong OSD.

Rook's CephCluster controller predicate ([predicate.go#L252-L255](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/controller/predicate.go#L252-L255)) is adjusted to trigger a reconcile when the annotation is added. The controller then runs cheap in-place checks:

1. **Confirmation matches.** Annotation value equals `yes-really-replace-osd-<id>` where `<id>` equals the Deployment's `ceph-osd-id` label.
2. **Target OSD exists and is not already destroyed.** `ceph osd dump`: the OSD must exist and its `state` must not contain `destroyed`.
3. **Deployment is host-based.** Label `ceph.rook.io/pvc` must be absent.
4. **Device references resolve now.** Best-effort check of the OSD's `spec.storage` references, per [Device references may not survive a swap](#device-references-may-not-survive-a-swap).

> On the first failed check the controller skips the OSD and emits a Warning event on the CephCluster CR; see [open question 2](#open-questions). On all checks passing, it sets the `ceph.rook.io/do-not-reconcile` label. From here the OSD and its Deployment are ignored by the controller's normal reconcile until the replacement completes or is cancelled, and the OSD health monitor goroutine takes over.

Reviewer: Is my assumption correct that this would be on the OSD deployment?

```suggestion
On the first failed check the controller skips the OSD and emits a Warning event on the CephCluster CR; see [open question 2](#open-questions). On all checks passing, it sets the `ceph.rook.io/do-not-reconcile` label on the OSD deployment. From here the OSD and its Deployment are ignored by the controller's normal reconcile until the replacement completes or is cancelled, and the OSD health monitor goroutine takes over.
```

- **Not rebuilding the destroyed OSD.** The prepare-job reports the OSDs it discovers with `ceph-volume list` to the status CM, today filtered only by cluster FSID ([volume.go#L1108-L1111](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/daemon/ceph/osd/volume.go#L1108-L1111)). The design extends that filter to also skip OSDs marked `destroyed` in the OSDMap, so the destroyed slot is never reported and the controller leaves it for the replacement.

Swapping the disk before the OSD is destroyed leaves no destroyed slot to pair with, so the blank disk is provisioned as a brand-new OSD with a new id. The `replace-ready-for-swap` annotation exists to make the correct ordering observable.

#### 5. Provision and complete

> When the user swaps the disk, the blocking Ceph signature — the FSID and other tags — disappears. On the next prepare-job reconcile, the prepare-job runs the existing discovery pass plus a new per-node pre-step:

Reviewer: I see this is already merged. A small detail here: I don't think Rook can detect the swap. Do we want to document for users that they can restart the operator pod, or is there another way we want them to kick off another reconcile? (perhaps a `swap-done` annotation)

This is a pretty small detail, and easily something that can be deferred for later

Author: agree. annotation is much better than operator restart. it also should work without restart if rook discovery daemon is enabled

    Ctrl->>Dep: recreate deployment<br/>from new OSDInfo (replicas=1)<br/>labels&annotations removed
    activate Dep
    Ctrl->>Ceph: confirm osd.5 up and in
    deactivate Dep
```

### Step-by-step

The walk-through uses the running example.

#### 1. Trigger and validation

The user sets `osd.rook.io/replace=yes-really-replace-osd-<id>` on the OSD Deployment. The numeric `<id>` in the value must equal the Deployment's `ceph-osd-id` label, a copy-paste guard against operating on the wrong OSD.

Rook's CephCluster controller predicate ([predicate.go#L252-L255](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/controller/predicate.go#L252-L255)) is adjusted to trigger a reconcile when the annotation is added. The controller then runs cheap in-place checks:

1. **Confirmation matches.** Annotation value equals `yes-really-replace-osd-<id>` where `<id>` equals the Deployment's `ceph-osd-id` label.
2. **Target OSD exists and is not already destroyed.** `ceph osd dump`: the OSD must exist and its `state` must not contain `destroyed`.

> 3. **Deployment is host-based.** Label `ceph.rook.io/pvc` must be absent.

Reviewer 2: (Optimization) If there are only PVC-based OSDs in the cluster, the replacement code path can be skipped.

Author: agree. but should we inform user that replacement will not happen? remove annotation, emit event, or log error?

Reviewer 2: I would suggest just to skip even checking for the annotation if we know it's a pvc-based cluster. The documentation seems sufficient.

The replacement logic is a state machine. Each tick the goroutine derives each OSD's phase from durable state — the OSDMap, the Deployment's labels and annotations, and the per-OSD cleanup Job — and takes the one action for that phase. It keeps no memory of its own, so an operator restart resumes wherever the state left it, and every action is idempotent.

| Phase | How the goroutine detects it | Action this tick |
| --- | --- | --- |
| Not validated | no `do-not-reconcile` label | ignore — the controller has not validated and labelled it yet |
| Not started | labelled; OSD still `in` | if no other OSD is draining and `ceph osd ok-to-stop <id>` passes, run `ceph osd out <id>` |
| Draining | labelled; OSD `out`; not yet `safe-to-destroy` | re-check `ceph osd safe-to-destroy <id>`, otherwise exit for retry |
| Ready to destroy | `safe-to-destroy` passes; slot not yet `destroyed` | run the [Destroy](#3-destroy) steps |
| Destroyed | slot `destroyed`; no `replace-ready-for-swap` annotation | add the `replace-ready-for-swap` annotation |
| Done | `replace-ready-for-swap` present | ignore — the goroutine's work is finished |

> Rook drains one OSD at a time. The drain is `ceph osd out <id>`, which keeps the daemon `up` while its PGs migrate to peers. For an OSD still `in`, the goroutine first runs `ceph osd ok-to-stop <id>` and marks it `out` only if the cluster can currently afford to lose it and no other OSD is draining. An OSD that is already `out` has passed that gate, so `ok-to-stop` is not run again — the goroutine only re-checks `ceph osd safe-to-destroy <id>` each tick until no PGs map to it. There is no timeout: a large OSD can take hours or days, and the user cancels by removing the annotation (see [Cancellation and retry](#cancellation-and-retry)).

Reviewer 2: Why not mark all annotated OSDs as `out`? A large cluster may frequently want to replace multiple OSDs at a time, and each OSD can take some time to drain. Even while some OSDs are in progress being replaced, more could be annotated for replacement and the goroutine would proceed with marking them out immediately. If we want a safeguard, maybe we only allow some percentage (e.g. 10%) of OSDs to be replaced at a time?

Author: `ok-to-stop` is the safeguard, applied per OSD: the goroutine checks one OSD, marks it out if it passes, and skips to the next if it doesn't. It reacts to actual PG placement, so a fixed percentage isn't needed.

Batching all annotated OSDs into one ok-to-stop is worse — they span different PGs, so the combined check fails more often and forces a guess about which to drop. cephadm does exactly that (drop half, retry) and its own comments call it rough: https://github.com/ceph/ceph/blob/0eceb0defba60152a8182f7bd87d164b639885b8/src/pybind/mgr/cephadm/services/osd.py#L484-L497

Reviewer 2: `ok-to-stop` is more strict of a check that we need. We are not actually stopping the OSD, we are only going to mark it as `out` in this step. The OSD will remain online while its PGs are draining. 

Let's take an example where the failure domain is `host`, where there are multiple hosts each with multiple OSDs. Now let's replace one OSD on multiple hosts. As long as there is spare capacity for one OSD being down on each host, all of the OSDs could be marked `out` without any negative side effects, and their replacement could happen simultaneously. However, `ok-to-stop` would mostly prevent OSDs from more than one host to be replaced at a time.

Ideally there would be a check for `ok-to-mark-out`, but of course that doesn't exist. My vote would be to not have any check before marking the OSDs `out`. If they too aggressively marked them for replacement and they see health issues, they could revert the replacement annotation and mark them back in, with only temporary impact. Or if extra safeguard is still desired, or at least make the `ok-to-stop` check optional.

Author: Agreed. So we drop the ok-to-stop gate, mark the annotated OSDs out, then poll safe-to-destroy before destroying each. Marking out stays reversible (ceph osd in) if the cluster shows stress.

Should we have some limit on how many osd we can drain at once? For example check percentage of out osds in cluster

Reviewer 2: We could have some percentage limit such as 10-20% at a time, but maybe such a limit is not necessary since it's reversible.

Author: agree, we can do it as follow up later and keep design simple for now.

Author: updated

This design proposes a workflow to replace a single failed OSD in place — preserving its OSD ID — without affecting other OSDs sharing the same metadata device.

## Notation

- **User** - a human cluster admin, or a third-party controller acting on their behalf, that edits Kubernetes objects.
- **Operator** - the Rook controller process (CephCluster controller).
- **Data LV / data device** - the LV (or block device) holding an OSD's bulk data. One per OSD.
- **DB LV / metadata device** - the LV holding the OSD's rocksdb (`block.db`). One per OSD; multiple OSDs can share the same metadata device.

## Goals and non-goals

Goals:

- Automate the single-OSD replacement process in a host-based cluster, preserving the OSD's config, CRUSH position, and DB LV (for shared-metadata OSDs), without interrupting the operator's normal reconcile flow or the other OSDs on the same node.
- Expose an API for the user to initiate a replacement and be notified when the disk may be safely swapped.

Non-goals:

> - **PVC-backed OSDs.** Issue #13240 is host-based storage, and PVC-backed OSDs run on a different code path; their replacement is left to a follow-up.

Reviewer 2: PVC-backed OSDs don't actually need this feature. They only use a unique metadata PV (device or partition), and don't share metadata devices across multiple OSDs.

### Device references may not survive a swap

Rook's `spec.storage` accepts non-persistent device references — `deviceFilter`, `useAllDevices`, kernel names, and udev symlinks under `/dev/disk/by-path`, `by-id`, and `by-uuid` — for both the data device and the `metadataDevice`. This is existing behavior and cannot be restricted without breaking backward compatibility. The design must tolerate it: it validates the affected OSD's references before drain as a best-effort check, and if a reference no longer resolves to the intended disk after the swap (for example, kernel names renumbered across a node reboot — see [persistent block device naming](https://wiki.archlinux.org/title/Persistent_block_device_naming)), provisioning fails with a clear error rather than provisioning the wrong disk.

## User story

A disk corresponding to `osd.5` fails on a node where five HDD OSDs share one NVMe metadata device. The user annotates the failed OSD's deployment:

```sh
kubectl -n rook-ceph annotate deployment rook-ceph-osd-5 \

>   rook.io/osd-replace=yes-really-replace-osd-5

Reviewer 2: How about this annotation? We are also adding `osd.rook.io` for another annotation in #17649.
```suggestion
  osd.rook.io/replace=yes-really-replace-osd-5
```

```

Rook drains and destroys `osd.5` in Ceph, scales its deployment to zero, and — once the OSD is destroyed and safe to pull — annotates the deployment `rook.io/osd-replace-ready-for-swap`. The user swaps the physical disk in the chassis at any later time, minutes or days, after watching for that annotation. Rook detects the new disk and brings `osd.5` back up with the preserved OSD ID 5. The other four OSDs on the same NVMe stay up the whole time.

## Proposed flow

The flow follows [Ceph's documented OSD-replacement procedure](https://docs.ceph.com/en/latest/rados/operations/add-or-rm-osds/#replacing-an-osd) (`ok-to-stop`, `osd out`, `safe-to-destroy`, `osd destroy`), the same one [cephadm](https://docs.ceph.com/en/latest/cephadm/services/osd/#replacing-an-osd) automates, in two phases:

> 1. **Destroy and reserve the slot.** The operator drains the OSD, then a short-lived Job destroys it in Ceph and performs any required host cleanup. The OSD Deployment is kept (scaled to zero) as the marker, and the operator signals via an annotation when the disk may be swapped.

Reviewer 2: ```suggestion
1. **Destroy and reserve the slot.** (Triggered by the `replace` annotation above) The operator drains the OSD, then a short-lived Job destroys it in Ceph and performs any required host cleanup. The OSD Deployment is kept (scaled to zero) as the marker, and the operator signals via an annotation when the disk may be swapped.
```

> 2. **Reprovision into the reused slot.** The existing prepare-job, with the adjustments described below, detects the swapped-in disk and provisions a new OSD into the destroyed slot, reusing its id and its surviving shared-metadata DB LV.

Reviewer 2: ```suggestion
2. **Provision the replaced OSD.** (Triggered by restarting the operator after the physical disk is replaced) The existing prepare-job, with the adjustments described below, detects the swapped-in disk and provisions a new OSD into the destroyed slot, reusing its id and its surviving shared-metadata DB LV.
```

Author: is it makes sense to add config for Cluster controller SyncPeriod, or mention Rook discovery daemon to replace OSD without manual restart?

Reviewer 2: If the discovery daemon is enabled, it can already trigger a reconcile when a new disk is detected. That should be a good approach to trigger the reconcile. I hope to avoid configuring a regular sync.

The destroy flow does not require the disk to have failed; it works uniformly for healthy OSDs too, since Rook is not responsible for hardware-failure detection.

### State and markers

The flow is multistep, and its state is carried by the following Ceph and Kubernetes data:

- **`rook.io/osd-replace` annotation on the OSD Deployment.** The caller's intent and a confirmation guard (value `yes-really-replace-osd-<id>`). Present from the trigger until the replacement finishes.

> - **The OSD Deployment, kept and scaled to `replicas=0`.** The operator never deletes it. It is kept for two reasons. First, a user can watch the replacement through the deployment and its annotations, without inspecting destroy Job state or running toolbox commands. Second, the deployment is itself the checkpoint: if the operator restarts mid-replacement, the annotated, scaled-down deployment is enough to resume, so no separate ConfigMap or status field is needed. The existing `ceph.rook.io/do-not-reconcile` label keeps Rook's normal reconcile from scaling it back up.

Reviewer 2: ```suggestion
- **The OSD Deployment, kept and scaled to `replicas=0`.** The operator only deletes it after replacement is fully completed. Before replacement, it is kept for two reasons. First, a user can watch the replacement through the deployment and its annotations, without inspecting destroy Job state or running toolbox commands. Second, the deployment is itself the checkpoint: if the operator restarts mid-replacement, the annotated, scaled-down deployment is enough to resume, so no separate ConfigMap or status field is needed. The existing `ceph.rook.io/do-not-reconcile` label keeps Rook's normal reconcile from scaling it back up.
```

Author: after replacement osd deployment will be recreated with the same name probably in the same reconclie loop.

Reviewer 2: Ok now I understand the doc. The OSD deployment really never is deleted since replacement just updates it.

> - **`rook.io/osd-replace-ready-for-swap` annotation.** Set by the operator once the OSD is destroyed. Its presence is the single contract signal that the disk may now be physically swapped; only the user consumes it — the operator does not use it.

Reviewer 2: ```suggestion
- **`osd.rook.io/replace-ready-for-swap` annotation.** Set by the operator once the OSD is destroyed. Its presence is the single contract signal that the disk may now be physically swapped; only the user consumes it — the operator does not use it.
```

    Op->>+J: next prepare-job reconcile on the node
    J->>Ceph: ceph osd tree --states destroyed (free slots on the node)
    J->>Ceph: ceph-volume lvm list 5 (surviving DB LV, or empty for single-disk)
    J->>Ceph: ceph-volume lvm prepare --osd-id 5 --data /dev/newDataDisk<br/>--block.db /dev/metaVG/survivingDbLV [--dmcrypt]
    Note over J: prepare-job writes OSDInfo to the per-node status CM
    J-->>-Op: succeeded

    Op->>Dep: clear replacement annotations and label,<br/>update spec, replicas=1
    activate Dep
    Op->>Ceph: poll ceph osd tree until osd.5 is up and in
    deactivate Dep
```

### Step-by-step

The walk-through uses the running example.

#### 1. Trigger

The user sets `rook.io/osd-replace=yes-really-replace-osd-<id>` on the OSD deployment. The numeric `<id>` in the value must equal the deployment's `ceph-osd-id` label (a copy-paste guard against operating on the wrong OSD).

A predicate carve-out on the CephCluster controller's owned-Deployment watch fires the reconcile when this annotation transitions from absent to present, value changes, or present to absent. Other Deployment updates remain suppressed.

#### 2. Validate

> Cheap upfront checks. On the first failed check the operator returns from the reconcile without requeuing; `ReportReconcileResult` emits a Warning event on the CephCluster CR. See [open question 2](#open-questions) on where to report the validation result.

Reviewer 2: Some of these checks seem unnecessary. For example, if the wrong annotation is used, it just won't match and nothing will happen. IMO we don't need any error for these conditions, we just need to state the conditions that are required, otherwise it won't be triggered. Or if it's a PVC-based cluster, the annotation would equally be ignored, no error necessary.

1. **Confirmation matches.** Annotation value equals `yes-really-replace-osd-<id>` where `<id>` equals the deployment's `ceph-osd-id` label.
2. **Target OSD exists and is not already destroyed.** `ceph osd dump`: the OSD must exist and its `state` must not contain `destroyed`.
3. **Deployment is host-based.** Label `ceph.rook.io/pvc` must be absent.
4. **Device references resolve now.** Best-effort check of the OSD's `spec.storage` references, per [Device references may not survive a swap](#device-references-may-not-survive-a-swap). A reference that does not resolve is reported; the flow still proceeds.
5. **Drain admission.** `ceph osd ok-to-stop <id>` must return OK. Failure means removing this OSD would push some PG below `min_size`; the user resolves the underlying cluster state before re-annotating.

On all checks passing, the controller proceeds to drain.

#### 3. Drain

Drain leaves the OSD `up` but with no PGs, then waits for `safe-to-destroy`.

1. **`ceph osd out <id>`.** Unconditional. Idempotent for already-out OSDs and required for `safe-to-destroy` to ever pass on a healthy OSD.
2. **Poll `ceph osd safe-to-destroy <id>` until OK.**

> Polling uses `RequeueAfter: 30s`, with a 1h cumulative timeout since drain start enforced via a wall-clock annotation `rook.io/osd-replace-started-at=<RFC3339>` stamped on the deployment on the first drain reconcile.

Reviewer 2: Why have a timeout? The drain may take several hours or even days depending on the cluster load and OSD size. If the admin decides they want to proceed, they can just swap the disk to move to the next step, right? 

edit: I thought this would be part of the job that you mention starts later in step 5. Waiting like this should not be done in the cluster reconcile IMO, as it would potentially impact/delay other reconciles.

Reviewer 2: Inside a job, seems like we just need a `for` loop and sleep every 30s until it's done, no need for requeue.

Reviewer 2: If they want to know when the job started, the job has a creation timestamp, no need for a new annotation.

Author: i agree that it will remove complexity from alredy big cluster controller but here we can properly handle cancellation in case if user will remove annotation. Waiting should not impact other reconciles because we can put replacement logic in the end of the cluster reconcile loop.

Author: is it makes sense to return something RequeueAfter: 1h from Cluster controller when we are waiting for osd replacement?

Reviewer 2: Having a long-term wait and requeue is still not ideal and it seems better to wait inside the job even for the cancel scenario. Let's continue on the new cancel discussion thread.

> Retry on failure or expiry: the user removes the `rook.io/osd-replace` annotation and re-adds it. The annotation-flip predicate fires a fresh reconcile; the operator clears the stale `started-at` annotation on a fresh start.

Reviewer 2: No need to re-add the annotation if the job just keeps running without a timeout.

#### 4. Stop the daemon

> Once `safe-to-destroy` passes, the operator stops the OSD daemon while keeping its Deployment. In a single Deployment update it sets the `ceph.rook.io/do-not-reconcile` label and `replicas=0` together in one atomic patch. The OSD updater already skips labeled deployments ([update.go#L155-L158](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/cluster/osd/update.go#L155-L158)).

Reviewer 2: Let's add the `do-not-reconcile` annotation when the job starts. I don't think we want any reconcile of the OSD once this process starts.

The operator then returns `RequeueAfter` and re-checks on a later reconcile rather than waiting for a Deployment event: the `do-not-reconcile` label makes the owned-object watch suppress this Deployment's update events ([predicate.go#L222-L224](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/controller/predicate.go#L222-L224)), so the pod's removal does not trigger a reconcile on its own. The destroy step proceeds only once the OSD pod is gone, because the daemon holds the dm-crypt mappings and LVs open while it runs.

#### 5. Destroy the OSD

> The operator creates a short-lived Replace Job to destroy the OSD and perform host cleanup. It reuses the prepare-job pod scaffold but has a separate entrypoint. All steps are idempotent on pod restart:

Reviewer 2: Oh, I was thinking the job was started with step 3. Seems like the job should be long-running, and its job is to drain and destroy the OSD. This offloads the burden of waiting for drain from the operator's reconcile loop.

Rook's cluster controller owns OSDs, so it also holds the replacement logic. The replacement step runs in the controller's top-level [`reconcile()`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/cluster/controller.go#L322), after [`reconcileCephCluster`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/cluster/controller.go#L458) has run the normal mon, mgr, and OSD reconcile, so replacement never preempts that work. It reuses the `ClusterInfo` and Ceph connection the controller already cached for that cluster, so nothing deeper in the reconcile chain changes signature.

Each waiting step (drain, daemon stop, destroy) returns `reconcile.Result{RequeueAfter}` instead of polling inline. Because `reconcile()` already returns a `reconcile.Result`, this is an ordinary successful reconcile — the CephCluster stays `Ready` and the controller-runtime queue handles re-entry. The controller reconciles one CephCluster at a time, which also keeps replacement serial with OSD provisioning and migration.

### Cancellation and retry

Cancellation is done by removing the `osd.rook.io/replace` annotation.

> - **Before destroy (during validate, drain, or after scale-to-zero).** Clean cancel: no Ceph state has been changed that the controller cannot reverse. The OSD has been marked `out`; removing the annotation makes the operator clear the `do-not-reconcile` label, the updater scales the Deployment back to `replicas=1`, and the user runs `ceph osd in <id>` to put it back into service. The operator clears the `started-at` annotation.

Reviewer 2: If the wait for `safe-to-destroy` is done inside the replace job, then cancellation seems simple. In addition to removing the replace annotation, the admin simply needs to delete the job and unset `out`. The running job would go away and stop waiting for the `safe-to-destroy`. 

If the job was past the `safe-to-destroy` then it would anyway already be completed and in the state of the next step where it is not supported to cancel the operation.

Reviewer 2: I still can't understand the concern with waiting in the job for `safe-to-destroy` instead of the operator.

Author: ok so in this case we encapsulate all "destroy" logic incuding downscaling OSD deployment into the Job. And Cluster controller responsibility will be doing some simple validation and creating the job.

Reviewer 2: Yes, that approach seems best, then we don't have any concerns about long reconciles, or requeue'ing, or if we need timeouts.

Author: moved logic to the job and updated the doc.

### Why keep the Deployment instead of deleting it

Deleting the OSD Deployment at teardown and recreating it after the swap would also work — the destroy half could checkpoint its state in a ConfigMap, as the migration flow does. The reason to keep the Deployment instead is that it is the natural place to expose the operation through the Kubernetes API: a user reads `kubectl get deployment` and its annotations to see that a replacement is in progress and when the disk is ready to pull, rather than inspecting Job state or running toolbox commands. Keeping it also makes it a free restart checkpoint — the annotated, zero-replica object survives an operator restart — so no separate ConfigMap is needed.

The cost is that a long-lived `replicas=0` OSD Deployment is a state Rook does not otherwise have — but Rook already has the `do-not-reconcile` label for a similar purpose.

> ### Why a Replace Job

Reviewer 2: Should it be called the "Prepare for Replace Job"? The replacing is not completed until later with the OSD prepare job after the physical disk swap.

- `nodes[].devices[].fullpath: "<value>"` — explicit DevLinks match (`/dev/disk/by-id/...`, `/dev/disk/by-path/...`). Compared against discovered symlinks, not regex.

Each shape interacts differently with the Linux device-naming interfaces:

- **Kernel names** (`vdb`, `sdc`, `/dev/sdc`) are assigned by the kernel at boot and [not guaranteed to be persistent](https://wiki.archlinux.org/title/Persistent_block_device_naming).
- **`/dev/disk/by-path/...`** is a udev symlink built from the sysfs port path: same physical port, same symlink.
- **`/dev/disk/by-id/...`** is a udev symlink built from the disk's hardware serial / WWN, unique per physical disk.
- **`/dev/disk/by-uuid/...`** is a udev symlink built from the filesystem or LV UUID, assigned at provisioning time.

The shapes that tolerate any swap (same-slot or different-slot, any new disk) are `useAllDevices` and `deviceFilter`. `by-path` tolerates only same-slot replacement. Kernel names tolerate only the lucky case where the kernel happens to assign the same name. `by-id`/`by-uuid` references in `name`/`fullpath` cannot work for a disk that hasn't been seen yet.

> The replacement flow must validate the affected OSD's CR references beforehand so the new disk is still resolvable under those references after the swap.

Reviewer 3: @arttor This statement, `OSD's CR`, is not clear to me. Could you explain to me? Thanks

Author: this wording is misleading. it is about osd references in Cluster CR.

status:
  phase: Replacing                             # Pending | Validating | Waiting | Replacing | Completed | Failed | Cancelled
  conditions:
    - type: Ready
      status: "False"
      reason: Replacing
      message: Replace Job in flight
      observedGeneration: 1
      lastTransitionTime: "2026-05-05T12:00:00Z"

  # captured at start of Replacing, before deployment delete
  osdInfo:
    node: node-1                                          # OSD deployment NodeSelector (operator reads via K8s API)
    dataLV: /dev/ceph-data-vg-5/osd-block-aaa...          # OSD deployment env ROOK_BLOCK_PATH
    dbLV: /dev/ceph-metadata-vg-1/osd-db-bbb...           # `[db].lv_path` from `ceph-volume lvm list` (Capture Job)
    crushDeviceClass: hdd                                 # OSD deployment env ROOK_OSD_CRUSH_DEVICE_CLASS
    encrypted: true                                       # `[db].tags.ceph.encrypted` from `ceph-volume lvm list` (Capture Job)
    osdFsid: 07bb0602-5e27-4fcc-86b1-c1faa0bc20ac         # `ceph osd dump`: `.osds[id=].uuid` (operator via mon)

  # populated on phase=Completed

>   newFsid: ""                                  # recorded on Completed

Reviewer 3: ```suggestion
  osdNewFsid: ""                                  # recorded on Completed
```

FSID might be confused with ceph cluster FSID, and also, we have `osdFsid` logged before. WDYT?

On all checks passing, advances to `Waiting`.

#### 3. Wait for replacement disk

Each reconcile, the controller checks whether the replacement disk is visible by reading the per-node status CM (`rook-ceph-osd-<node>-status`, populated by the prepare-job). If the empty replacement disk isn't there yet, requeue. When it appears, advance to `Replacing`.

Cancel during `Waiting` is clean: no Ceph or host state has been changed, no LVs touched. Deletion of the CR ends the flow with no recovery needed.

If `spec.diskWaitTimeout` (default 24h) is exceeded, transitions to `Failed` with `reason=ReplacementDiskMissing`. After timeout, the user can insert the disk and create a new CR for the same OSD ID (the OSD is still alive in Ceph at this point), or delete the CR.

#### 4. Replace

> `Replacing` runs the full set of state changes in sequence. On each reconcile, the operator inspects observable state (deployment presence, `status.osdInfo`, Replace Job status, `ceph osd tree`) and runs the next unfinished sub-step.

Reviewer 4: The Capture Job reads ceph volume lvm list from the target node assuming the NVMe metadata device is still healthy. If the failure is a controller failure that takes out the NVMe itself, the Capture Job fails with no recovery path. The design should either allow dbLV to be specified manually in the spec, or document clearly that this scenario is out of scope.

Author: good point. so validate step should check if used metadata is healthy and fail replacement early. 

maybe we can check health of other OSDs sharing the same metadata device or run capture job on validation phase.

  --block.db <dbLV> \
  --crush-device-class <crushDeviceClass>
```

   The Job writes the new OSD's info to `rook-ceph-osd-<node>-status` (the per-node CM Rook already uses to drive daemon creation).

6. **Create new Deployment.** The cluster controller's existing path takes over: `createOSDsForStatusMap` ([`status.go#L324`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/cluster/osd/status.go#L324)) reads the per-node status CM the Replace Job wrote and creates the daemon Deployment.

7. **Wait for `up+in`.** The controller polls `ceph osd tree` each reconcile until the new daemon is `up` AND `in`. Once visible, capture `osd_uuid` from `ceph osd metadata <id>` and transition to [Complete](#5-complete).

#### 5. Complete

Terminal phase. `.status.newFsid` and `.status.completedAt` are recorded; the `Ready` condition transitions to `True`.

### Cancellation

Cancel = delete the `CephOSDReplace` CR; a finalizer runs any cleanup needed.

**Pending, Validating, Waiting.** Clean cancel — no Ceph or host state has been changed. The finalizer is a no-op aside from clearing the auto-provisioning gate label on the affected node.

> **Replacing — best-effort, deferred.** Once `Replacing` begins, the operator commits to running through to a terminal phase. Cancel intent is recorded but not acted on mid-flow:

Reviewer 4: If lvm prepare hangs (bad replacement disk, I/O errors), the Replace Job never reaches a terminal state, and the ROOK_DISCOVER_ONLY gate stays open on that node forever, blocking all new OSD  provisioning. The Job needs a wall-clock timeout.

Each shape interacts differently with the Linux device-naming interfaces:

- **Kernel names** (`vdb`, `sdc`, `/dev/sdc`) are assigned by the kernel at boot and [not guaranteed to be persistent](https://wiki.archlinux.org/title/Persistent_block_device_naming).
- **`/dev/disk/by-path/...`** is a udev symlink built from the sysfs port path: same physical port, same symlink.
- **`/dev/disk/by-id/...`** is a udev symlink built from the disk's hardware serial / WWN, unique per physical disk.
- **`/dev/disk/by-uuid/...`** is a udev symlink built from the filesystem or LV UUID, assigned at provisioning time.

The shapes that tolerate any swap (same-slot or different-slot, any new disk) are `useAllDevices` and `deviceFilter`. `by-path` tolerates only same-slot replacement. Kernel names tolerate only the lucky case where the kernel happens to assign the same name. `by-id`/`by-uuid` references in `name`/`fullpath` cannot work for a disk that hasn't been seen yet.

The replacement flow must validate the affected OSD's CR references beforehand so the new disk is still resolvable under those references after the swap.

## Current gaps

> Rook has no automated flow for replacing a failed OSD today. The closest existing primitive is the migration flow (`spec.storage.migration`), which recreates OSDs in place after encryption or store-type spec changes: it destroys the OSD and re-prepares with `ceph-volume raw prepare --osd-id` via the `ROOK_REPLACE_OSD` env var. Migration only covers raw-mode OSDs; the shared-metadata case needs the following fixes:

Reviewer 5: migration of OSD is only supported and tested for OSDs on PVC.

3. **No discover-only mode in the prepare-job.** The prepare-job combines discovery and provisioning in a single sequential pass (`Provision`, [daemon.go#L159-L283](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/daemon/ceph/osd/daemon.go#L159-L283)) — no way to inventory a node without auto-claiming any empty disks it finds. The replacement flow needs "scan but don't claim" so the empty replacement disk doesn't get auto-provisioned with a fresh ID before the operator can drive `ceph-volume lvm prepare --osd-id`. This design adds a `ROOK_DISCOVER_ONLY` mode to the prepare-job; the cluster controller passes it for nodes with an active `CephOSDReplace`.

## Proposed flow

> This flow orchestrates [Ceph's documented OSD-replacement procedure](https://docs.ceph.com/en/latest/rados/operations/add-or-rm-osds/#replacing-an-osd) (`safe-to-destroy` → `osd destroy` → `lvm zap` → `lvm prepare --osd-id` → `lvm activate`) inside a single short-lived Kubernetes Job, with state machine maintained in Rook CR status. `cephadm` — Ceph's container-orchestrator analogue — preserves OSD IDs by default ([cephadm OSD service docs](https://docs.ceph.com/en/latest/cephadm/services/osd/#replacing-an-osd)); this design follows the same convention.

Reviewer 5: All data devices must be passed to ceph-volume lvm batch, even if only one device is to be replaced. For example, if `deviceA` was replaced and rook is trying to `prepare` it again, then other data devices must be passed in as well.  Currently the new CR does not have have that information.  How do we plan to address that? 

```
ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>
```

Author: good catch. But the design intentionally uses `lvm prepare` (not `lvm batch`) for exactly this reason. The destroyed OSD's databaseSizeMB, source metadataVG, and data-LV path are captured into `.status.osdInfo` before the deployment is deleted. The Replace Job then lvcreates a new DB LV at the captured size inside the original metadata VG and runs 
`lvm prepare --data <new-disk> --block.db <new-db-lv> --osd-id <id>` with explicit paths - no sibling data devices need to be passed because nothing on them is touched.

I tested this flow end-to-end on a 2-OSD shared-metadata setup on VM: osd.1 was destroyed and rebuilt with the same ID on a fresh data disk, the sibling osd.0 (sharing the same metadata device) stayed up and the cluster stayed HEALTH_OK the whole time.

    Op->>Ceph: safe-to-destroy 5 (poll until OK)
    Op->>OldPod: read deployment env
    Op->>+J: create Capture Job (env: osdId)
    J->>J: ceph-volume lvm list --format json
    J->>J: write captured fields to per-replacement CM
    J-->>-Op: Succeeded
    Op->>Ceph: ceph osd dump (osdFsid)
    Op->>CR: capture OSD info
    Op->>OldPod: delete deployment
    destroy OldPod
    Op->>OldPod: wait for pod termination
    Op->>+J: create Replace Job (env from osdInfo)
    J->>Ceph: safe-to-destroy 5 (re-check)
    J->>Ceph: ceph osd destroy osd.5
    J->>Ceph: config-key rm dm-crypt key (if encrypted)
    J->>J: cryptsetup close db mapping (if encrypted)
    J->>J: lvremove db lv
    J->>J: ceph-volume lvm zap data lv
    J->>J: lvcreate new db lv
    J->>Ceph: ceph-volume lvm prepare --osd-id 5
    J->>J: write new OSD info to per-node status CM
    J-->>-Op: Succeeded
    create participant NewPod as New OSD pod
    Op->>NewPod: create deployment (id=5, dataLV, new dbLV, encrypted)
    NewPod->>Ceph: join cluster
    Op->>Ceph: ceph osd metadata 5 until Ready
    Op->>CR: phase=Completed
```

### Open question: controller placement

> The diagram doesn't pick a concrete CR or controller for the replacement reconcile logic. Two candidates: extend the existing CephCluster controller (which already hosts `spec.storage.migration`), or introduce a separate `CephOSDReplace` CRD with its own controller. The design leans toward the separate CRD for the following reasons:

Reviewer: FWIW, I also lean toward a separate CRD for the additional reason that (IMO) it goes against normal declarative management designs to keep historical info about disk replacements on the CephCluster. That info isn't really relevant if someone were to want to use the same CephCluster spec in a different k8s cluster.

I think it is best if historical records of replacements can be purged from the current cluster if desired -- or omitted from copy-pasted clusters -- without any special handling and without replacement info being present.

Ideally, OSD replacement is just an imperative operation with no CRD, as with our OSD purge job. But because of the complexity of the end-to-end flow for OSDs with metadata, I think having a CRD is a reasonable way to tackle the problem.

Author: maybe we can aloso annotate failed OSD pod scheduled for replacement. In this case no replacement history in CRDs. But reconcile replacement logic will be in CephCluster controller.

Reviewer 4: This creates a requirement for ops teams to kubectl create CephOSDReplace introduces a K8s RBAC requirement that is often not acceptable in hardened clusters, the person physically swapping a disk typically shouldn't need custom  resource create/update permissions in the rook-ceph namespace. The design itself acknowledges spec.storage.replaceOSD on CephCluster (mirroring spec.storage.migration) as a fallback but dismisses it.
I agree it is not a general use case, but this is a valid point for most big organizations. Given the operational reality, the CephCluster field approach should be the primary path, not the fallback.

> 1. **CephCluster's `reconcileCephDaemons` is monolithic and synchronous** — mon, mgr, and osd reconcile run sequentially in one call ([`cluster.go#L116-L160`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/operator/ceph/cluster/cluster.go#L116-L160)); `osd.Cluster.Start()` returns plain `error`, so there's no way to express terminal failure (bad CR rejected) vs. transient `RequeueAfter` (waiting for disk-swap or Job completion). Adding long-running multi-step logic to this path interferes with mon/mgr reconcile and lacks the return semantics the flow needs.

Reviewer: Something I am considering is the possibility for race conditions between this controller and the CephCluster controller. 

Even if we don't think it's truly needed, I think it's safest to use locking and blocking to ensure the CephCluster controller and OSD replacement controller can't operate simultaneously on OSDs. 

I think we could do 2 things to ensure this:
1. apply a "do not reconcile" annotation/label to the OSD pod being replaced to indicate to the CephCluster controller that it shouldn't reconcile the replacing OSD
2. use an OSD reconciliation mutex lock that prevents the two controllers from managing OSDs at the same time

This could result in a bit more system latency when OSDs are being worked on, but it would ensure we don't have unexpected race conditions. IMO, a worthwhile tradeoff.

Reviewer 6: What if the `CephCluster` controller is blocked from any modification to a OSD if a ` CephOSDReplace` CR exists for the OSD and has a `.status.phase` of something other than complete/finished/done?

Author: > What if the `CephCluster` controller is blocked from any modification to a OSD if a ` CephOSDReplace` CR exists for the OSD and has a `.status.phase` of something other than complete/finished/done?

good idea. to make this desgin work, cluster controller will have to lookup non-reminal OSDs to block OSD provisioning on the node. we can also use this check to skip OSD reconcile

2. **Replacement state has to survive between reconciles**, and the cluster controller has no existing place to store sub-operation state — adding one (a side ConfigMap, or extending `CephCluster.status`) is part of the cost.

Concrete shape of each candidate:

> - **Extend the cluster controller** — state in either a side ConfigMap (`osd-replacement-state`, similar to `osd-migration-config`) or `CephCluster.status`.

Reviewer 6: What is the motivation to use a separate CM?  Is there concern that the OSD state would be large enough to start bumping into etcd limits?

Author: Mostly because existing workflow of osd-migration is using CM. Cluster CR status can also be used.

- **New `CephOSDReplace` CRD + dedicated controller** — state on `.status`. Independent reconcile loop; doesn't modify the existing OSD path (shares only the pod-scaffold builder, `provisionPodTemplateSpec`). Light coupling on the cluster side: skip auto-provisioning on affected nodes.

The rest of this design is based on a separate `CephOSDReplace` CRD, with implications for the cluster-CR fallback flagged inline.

### CRD proposal

Config lives on `CephOSDReplace.spec` and state in `.status`. `spec.cephCluster` and `spec.osdId` are immutable post-create. `.status` carries phase and conditions following the K8s operator pattern.

```yaml
apiVersion: ceph.rook.io/v1
kind: CephOSDReplace
metadata:
  name: replace-osd-5
  namespace: rook-ceph
  labels:

>     rook.io/osd-replacement-node: node-1       # operator-managed; equals the target OSD's host node

Reviewer: If this is operator-managed, it would probably be best as an annotation.

Reviewer 6: If the OSD ID immutable, would it make sense to make the node name immutable as well?  I assume the intent is to provide a convenience for the user by looking up the node name for them from the OSD ID. However, if the user already had to provide device paths from the node, I'm not sure how much effort that saves.

Or we could go one step further, is the OSD ID alone enough information to infer the LV paths?

Author: > If this is operator-managed, it would probably be best as an annotation.

Correct, it is managed by operator to lookup replacements by node. Agree, annotation should be used instead 

> Or we could go one step further, is the OSD ID alone enough information to infer the LV paths?

yes, osd-id+clusterID should be enought to identify OSD and collect the rest from the cluster.

> spec:

Reviewer: I would like to consider adding a "status" field into the `spec` here. Normally, this would be inadvisable, but I think we have special considerations here. I'd appreciate @travisn 's and @jhoblitt 's thoughts on this as well.

Many backup tools only back up a portion of metadata and specs, and they leave the status out. Because we are tracking an imperative operation with this declarative spec, there is risk that a user naively copy-pasting resources from K8s cluster A to cluster B might end up in a situation where CephCluster is created in B, and then immediately OSDs start being replaced. If we keep a field (I'll tentatively call it `replacementStatus: Completed`) in the spec that is monotonic, the copy on B can see that replacement is done and take no action. A CEL rule on the CRD can ensure `replacementStatus` cannot change back from `Completed` once it reaches that state.

It seems reasonable to me to track all other status in the actual `status`.

Reviewer 6: Is there any precedent for tracking state under .spec?  I can't think of an example.  As an example,  `Job` uses `.status.succeeded`

Author: controller will wait until new disk is available in cluster invetory. If there is no candidate, replacement process will not be started and old OSD will not be destroyed. But if the disk was available before this CR was created, the disk should be already discovered and provisioned by Rook as new OSD. So accidental copy should not lead to unexpected osd provisioning or destruction if it was noticed and deleted before disk was physically replaced.

Reviewer: PVCs track state under .spec. `.spec.volumeName` is ~essentially~ a status for dynamic provisioning, and AFAIU from talking to sig-storage, that ensures the PV binding info isn't lost if resources are ported naively.

  cephCluster: my-cluster                      # immutable; target cluster in this namespace
  osdId: 5                                     # immutable
  confirmation: yes-really-replace-osd-5       # must equal "yes-really-replace-osd-{osdId}"; copy-paste guard against operating on the wrong OSD

>   autoOut: false                               # optional; if true, operator marks healthy OSD `out` automatically (during Replacing). Default: false (fail-fast on up+in at Validating)

Reviewer 6: My head exploded. 💥 

Doesn't the OSD need to be purged from ceph?  Even though the ID is recycled, isn't it functionally a new OSD being added to the cluster?

Reviewer 6: If the OSD isn't purged, would ceph keep asserting warnings for the "old" OSD. E.g. slow op warnings

Author: Defatul flow assumes that OSD for replacement is not healtahy (pod is down). So if i got it correctly,  osd will be marked as out by ceph and after data will be rebalanced, safe-to-destroy check will retrun OK.

I added this option (not sure if it needed) if user wants for some reason to replace healthy osd. Maybe it is better to remove this option from CR and ask user to out OSD manually in this case.

Reviewer: I can imagine instances of healthy replacement. Perhaps disk SMART stats indicate disk failure is likely soon, or perhaps a user is phasing out old disks cluster-wide.

Reviewer 2: I believe the ceph term is that the OSD would be marked as `destroyed`. This will leave the OSD ID intact, but the OSD ID will be available for provisioning a new OSD in its place.

  safeToDestroyTimeout: 1h                     # optional; how long Replacing tolerates EBUSY on safe-to-destroy before Failed. Default: 1h
  diskWaitTimeout: 24h                         # optional; how long Waiting tolerates a missing disk before Failed. Default: 24h

status:
  phase: Replacing                             # Pending | Validating | Waiting | Replacing | Completed | Failed | Cancelled
  conditions:
    - type: Ready
      status: "False"
      reason: Replacing
      message: Replace Job in flight
      observedGeneration: 1
      lastTransitionTime: "2026-05-05T12:00:00Z"

  # captured at start of Replacing, before deployment delete
  osdInfo:
    node: node-1                                          # OSD deployment NodeSelector (operator reads via K8s API)
    dataLV: /dev/ceph-data-vg-5/osd-block-aaa...          # OSD deployment env ROOK_BLOCK_PATH
    dbLV: /dev/ceph-metadata-vg-1/osd-db-bbb...           # `[db].lv_path` from `ceph-volume lvm list` (Capture Job)
    metadataSourceDevice: /dev/vdd                        # `[db].devices[0]` from `ceph-volume lvm list` (Capture Job)

>     metadataVG: ceph-metadata-vg-1                        # `[db].vg_name` from `ceph-volume lvm list` (Capture Job)

Reviewer 6: If the metadata device lv path is provided, are the PV and VG redundant information?

Author: yes, it looks like it is redundant and probabaly not needed if lvm zap will be used. If i got it correctly, it will need only `dbLV`

>     crushDeviceClass: hdd                                 # OSD deployment env ROOK_OSD_CRUSH_DEVICE_CLASS

Reviewer 6: I assume this is to support replacing an `nvme` with an `hdd`?  What happens if this value conflicts with an explicit declaration in the `CephCluster.spec.storage`?

Author: The opposite, actually. it's captured to preserve the existing OSD's crush class across replacement. It's read from the destroyed OSD's deployment env ROOK_OSD_CRUSH_DEVICE_CLASS.

>     databaseSizeMB: 1500                                  # `[db].lv_size` (bytes) / 1048576 from `ceph-volume lvm list` (Capture Job)

Reviewer 6: Why does this need to be explicitly set?  What happens if this value doesn't match the existing LV?

Author: I think it we can reuse value from cluster spec instead. 
I decided to make it explicit and capture value directly from destroyed OSD to use what the existing OSD was actually allocated, not whatever the spec says now (which may have drifted).

Author: and with your suggested zap approach, this field is probably not needed.

>     encrypted: true                                       # `[db].tags.ceph.encrypted` from `ceph-volume lvm list` (Capture Job)

Reviewer 6: What happens if this value conflicts with an explicit declaration in the CephCluster.spec.storage?

Reviewer 6: Is the goal to support transitioning an OSD between encrypted <-> no encrypted as part of replacement?

Author: no, status stores options from old destroyed OSD to create new osds with the same options. In general, this info is not required and we can read ClusterCR but i thought that explicit copy of params will be more maintainable.

>     osdFsid: 07bb0602-5e27-4fcc-86b1-c1faa0bc20ac         # `ceph osd dump`: `.osds[id=].uuid` (operator via mon)

Reviewer 6: Does the fsid persist in the mon if the OSD is down?

Author: yes

  # populated on phase=Completed
  newFsid: ""                                  # recorded on Completed
  completedAt: null
```

Users cancel a replacement by deleting the `CephOSDReplace` CR. A finalizer gives the operator a chance to clean up before the CR is removed. If the user cancels before the operator picks up the swapped disk, no Ceph or host state has changed and the CR is removed cleanly. If the replacement has already started, the operator runs it to a terminal state (success or failure) before removing the CR — see [Cancellation](#cancellation).

CR names are arbitrary. To re-replace the same OSD, the user creates a new CR with a different name. Terminal CRs (`Completed`, `Cancelled`, `Failed`) for the same `osdId` are ignored by the operator and can be deleted when no longer useful.

#### Coordination

> Replacements run serially per CephCluster as a simplifying choice, matching cephadm's `osd rm` queue and Rook's existing OSD migration. Per-OSD `safe-to-destroy` only returns OK once the OSD is fully drained from every PG's acting set, so concurrent destroys of independently-safe OSDs are technically safe — but serial keeps the operational model simple.

Reviewer 6: 👍

> The queue is implemented via a `Pending` phase. Each reconcile, the controller lists peer `CephOSDReplace` CRs in the same namespace targeting the same cluster. If no earlier-`creationTimestamp` peer is in a non-terminal phase, this CR advances to `Validating`; otherwise it stays in `Pending` and re-checks next reconcile. UID breaks same-second ties.

Reviewer: Excellent points. +1

> Extending CephCluster with a `spec.storage.replaceOSD` field needs no coordination logic — a single field admits only one in-flight replacement.

#### Auto-provisioning skip

> The Rook cluster controller spawns the prepare-job, which by default auto-discovers devices and provisions new OSDs. To make the replacement flow work, the cluster controller must run the prepare-job in "discover only" mode on a node where a replacement is running — discovery happens, provisioning doesn't.

Reviewer: I think I understand the reason for this, but I'd appreciate confirmation or clarification. I think it would also be good to clarify in the doc.

I think I understand that this is needed because if the prepare job were to run after replacement begins, it might pick up a newly-inserted disk and provision it, but it wouldn't be properly set up with the reused metadata partition? Is that right?

Reviewer 6: Is there a race condition where provisioning is blocked on the node where OSD replacement is occurring, that provisioning occurring on another node could "steal" the OSD id?

Author: > I think I understand that this is needed because if the prepare job were to run after replacement begins, it might pick up a newly-inserted disk and provision it, but it wouldn't be properly set up with the reused metadata partition? Is that right?

yes, based on my understanding of the current workflow, operator will:
- spawn prepare job on every node to discover new OSDs
- the job will automatically start OSD provisioning with new IDs in the same step. 

for raw OSD it is not the biggest problem even though ceph docs suggest to reuse OSD ID for replacement. But for shared metadata device this flow will not work. Operator should clean up metatada slot from prev OSD and use it in prepare job for replacement.

> Is there a race condition where provisioning is blocked on the node where OSD replacement is occurring, that provisioning occurring on another node could "steal" the OSD id?

good question. you mean if on one node we destroyed old osd and someone added new osd to different node? Probably, in this case it will be better to block provisioning on all nodes.

The walk-through uses the running example.

#### 1. Trigger — user creates a `CephOSDReplace` CR

Typical case is a failed disk: Ceph auto-marks the OSD `down` and `out` and rebalances the data. User creates a `CephOSDReplace` CR and replaces the failed disk in the datacenter.

Healthy (`up+in`) OSDs are rejected by [Validate](#2-validate) unless the user marks the OSD out manually first (`ceph osd out <id>`) or sets `spec.autoOut: true`.

On creation, the CR enters `Pending` and waits for any earlier in-flight replacement to terminate. Once cleared, it advances to `Validating`.

#### 2. Validate

Cheap upfront checks. Each reconcile cycle runs the checks in order; the first failure ends the phase.

1. **Confirmation matches.** `spec.confirmation` must equal `"yes-really-replace-osd-{spec.osdId}"`. On mismatch: `Failed` with `reason=InvalidSpec` (typo guard).
2. **Target OSD exists.** If absent from the OSD map: `Failed` with `reason=InvalidSpec`.
3. **Target OSD is destroyable.** If `up && in` and `spec.autoOut: false`: `Failed` with `reason=OSDStillIn`. If `up && in` and `spec.autoOut: true`: accepted; the actual `ceph osd out` runs in [Replace](#4-replace), not here.

> 4. **CR-level device matching is swap-tolerant.** The OSD's data device must be referenced via `useAllDevices`, `deviceFilter`, or a `/dev/disk/by-path/...` path (same-slot replacement only). Kernel names (`vdb`, `/dev/sda`), `by-id`, and `by-uuid` references are rejected — they can't resolve to a fresh disk. On rejection: `Failed` with `reason=InvalidSpec`. (Whether to make this configurable is [open question 6](#open-questions).)

Reviewer 6: I often specify nvmes by `/dev/disk/by-id`, as I have been bitten by firmware updates changing the pci bus enumeration order.  This could be important in hosts which have multiple devices of the same model in the chassis for different uses, so a simple deviceFilter regex match isn't possible.

What if the drive bay has died and the replacement drive must be at a different path?  Replacement would only be possible with `useAllDevices` or `deviceFilter`?

Author: > What if the drive bay has died and the replacement drive must be at a different path? Replacement would only be possible with useAllDevices or deviceFilter?

yes. In general, i was not sure if we need to validate this because of the amount of edge cases. But migrating osd config in CR is a real problem

Each reconcile, the controller checks whether the replacement disk is visible by reading the per-node status CM (`rook-ceph-osd-<node>-status`, populated by the prepare-job). If the empty replacement disk isn't there yet, requeue. When it appears, advance to `Replacing`.

Cancel during `Waiting` is clean: no Ceph or host state has been changed, no LVs touched. Deletion of the CR ends the flow with no recovery needed.

If `spec.diskWaitTimeout` (default 24h) is exceeded, transitions to `Failed` with `reason=ReplacementDiskMissing`. After timeout, the user can insert the disk and create a new CR for the same OSD ID (the OSD is still alive in Ceph at this point), or delete the CR.

#### 4. Replace

`Replacing` runs the full set of state changes in sequence. On each reconcile, the operator inspects observable state (deployment presence, `status.osdInfo`, Replace Job status, `ceph osd tree`) and runs the next unfinished sub-step.

1. **autoOut (conditional).** If the OSD is `up && in` and `spec.autoOut: true`, run `ceph osd out <id>`. For the typical failed-disk case the OSD is already `out` (Ceph auto-marked it after `mon_osd_down_out_interval`) and this step is a no-op.

> 2. **Wait for `safe-to-destroy` OK.** `ceph osd safe-to-destroy <id>` returns OK only after the OSD is fully drained from every PG's acting set. Requeued until OK. If `spec.safeToDestroyTimeout` (default 1h) is exceeded, transitions to `Failed` with `reason=NotSafeToDestroy`.

Reviewer 5: The `safe-to-destroy` check could be an operator side operation that is run as a pre-requisite before starting the job. Else the `replace` job might hold a slot, on the node,  for potentially hours while it polls `ceph osd safe-to-destroy` in a loop.

5. **Replace Job.** Built on the existing prepare-job pattern (same one the OSD migration flow uses): a new rook subcommand on the prepare-job's pod scaffold (`provisionPodTemplateSpec`) runs the destroy + prepare bash sequence. The new DB LV's UUID is generated by the operator at Job creation and passed via env.

   The container's command:

```bash
set -euo pipefail

# Skip safe-to-destroy + destroy if osd.5 is already destroyed (idempotent on Job retry).
already_destroyed() {
  ceph osd dump --format json | python3 -c "
import sys, json
osds = json.load(sys.stdin).get('osds', [])
o = next((o for o in osds if o.get('osd') == 5), None)
sys.exit(0 if o and 'destroyed' in (o.get('state') or []) else 1)
"
}

if ! already_destroyed; then
  # Re-check safe-to-destroy (insurance against races between phase transition and Job start).
  ceph osd safe-to-destroy 5
  # Destroy in Ceph (preserves OSD ID 5 for reuse).
  ceph osd destroy osd.5 --yes-i-really-mean-it
fi

> # Remove dm-crypt key (no-op on Ceph v19+; defensive for older versions).

Reviewer 6: If we assume that this functionality will not be backported to Rook 1.18.x, the proposal only needs to consider support for Squid or newer.

ceph config-key exists dm-crypt/osd/<osdFsid>/luks \
  && ceph config-key rm dm-crypt/osd/<osdFsid>/luks

# Close DB-side LUKS mapping.
DB_MAPPING=$(lsblk -nlo NAME,TYPE /dev/ceph-metadata-vg-1/osd-db-<dbUuid> | awk '$2=="crypt"{print $1; exit}')
[ -n "$DB_MAPPING" ] && cryptsetup status "$DB_MAPPING" >/dev/null 2>&1 \
  && cryptsetup close "$DB_MAPPING"

# Free the DB slot.
lvs /dev/ceph-metadata-vg-1/osd-db-<dbUuid> >/dev/null 2>&1 \

>   && lvremove -f /dev/ceph-metadata-vg-1/osd-db-<dbUuid>

Reviewer 6: Does the metadata lv need to be deleted and recreated? Or could it be zapped and reused?

Author: Honestly, I went with lvremove + lvcreate-with-fresh-UUID for retry idempotency. but I wasn't aware lvm zap (without --destroy). it looks like it does exactly what is needed here and can collapse 3 commands of proposed replace job.
 I'll switch the design to zap+reuse.

Author: Done. tested on VM - zap works. Changed the doc accrodingly.

- On Job failure, the finalizer spawns a one-shot cleanup Job on the target node (same pod scaffold) to remove any partially-allocated DB LV before removing the CR.
- On Job success, cancel is not honored — the new OSD joins the cluster.

## Notes on Scope

### Multiple metadata devices on one node — works conditionally

Rook supports per-device metadata-device pairing:

```yaml
nodes:
- name: "node-1"
  devices:
  - name: "/dev/disk/by-path/...sda"
    config: { metadataDevice: "nvme0n1" }
  - name: "/dev/disk/by-path/...sdb"
    config: { metadataDevice: "nvme0n1" }
  - name: "/dev/disk/by-path/...sdc"
    config: { metadataDevice: "nvme1n1" }   # different metadata device on the same node
```

This setup requires exact `name:` (or `fullpath:`) references — the per-device `config:` block can only be attached to a specific device entry, not to a regex match. Replacement of a single OSD on this setup works structurally (each OSD's `metadataSourceDevice` is captured in its `osdInfo` at destroy time), with two caveats:

- **Device-name validation must permit exact entries** — see [open question 6](#open-questions).

> - **Same-slot replacement is required** — `by-path` resolves only when the new disk is in the original slot. Different-slot replacement stalls in the Wait step.

Reviewer 6: This would be an unfortunate restriction as I've experienced a lot of drive bay failures over the years (SAS backplanes hate me).

### PVC-based OSD replacement — separate design

PVC-backed OSDs use a different code path (raw mode via `GetCephVolumeRawOSDs`, separate destroy plumbing). Issue #13240 is host-based storage; PVC replacement is a separate design.

## Open questions

1. **Controller placement.** Design leans toward a separate `CephOSDReplace` CRD; `spec.storage.replaceOSD` on CephCluster (mirroring `spec.storage.migration`) is a fallback — see [Open question: controller placement](#open-question-controller-placement). Maintainers' call.

> 2. **Parallelism.** The proposed OSD replacement process is serial. Are there use-cases for parallel replacement we should support — multiple OSDs safe-to-destroy on the same node, all safe-to-destroy in the cluster at once, configurable concurrency?

Reviewer 6: I believe concurrency would be purely a performance optimization rather than allowing any different functionality?

Author: yes

Reviewer 5: This might be an operational complexity. Concurrent Jobs on the same node touching the same metadata VG, LV  might create race conditions and make debugging harder.

Reviewer 4: Two disks failing on different nodes simultaneously means the second replacement queues behind the first one, including its safe-to-destroy drain period. Serial-per-node would be a  better default, no shared resources contend between two replacements on separate nodes

> 3. **Auto-replace mode.** The proposed flow is always triggered explicitly by the user (CR creation). Should there be a follow-up option for automated replacement that triggers the same flow when a failed OSD and a fresh disk are detected on a node?

Reviewer 6: I suppose a related feature would be an explicitly designated hot spare.

Reviewer 6: (I'm not proposing that as a feature -- just mentioning that it would be a similar workflow)

> 3. **Auto-replace mode.** The proposed flow is always triggered explicitly by the user (CR creation). Should there be a follow-up option for automated replacement that triggers the same flow when a failed OSD and a fresh disk are detected on a node?

Reviewer 5: Rook might not be able to distinguish a `failed disk replaced by admin` from `transient I/O error on a healthy disk` or  `disk removed for decommissioning`.  Auto-triggering a destructive osd destroy on a false positive would be catastrophic.

> 4. **Default values.** Proposed: `safeToDestroyTimeout: 1h`, `diskWaitTimeout: 24h`, disk-wait re-check interval `5 min` (cluster-config tunable). Reasonable, or do reviewers see a reason to change them?

Reviewer 6: An hour seems like a fairly short default. Migrating PGs off a 20+ TB HDD could take a week or more.

5. **Disk-swap responsiveness.** Can the design rely on Rook's existing discovery (rook-discover when enabled, otherwise the per-reconcile prepare-job inventory) to detect the replacement disk?

> 6. **Device-name validation.** Proposed: reject kernel names (`/dev/sda`, `vda`), `by-id`, and `by-uuid` references; accept `useAllDevices`, `deviceFilter`, and `by-path` (with implicit same-slot expectation). Sample:

Reviewer 6: I don't think it is critical that replacing drives with a different path is part of the first iteration but I'm fairly sure replacing a drive in different slot or replacing an existing by-id device would be a common use case.

   ```yaml
   spec:
     storage:
       nodes:
       - name: node-1
         devices:
         - name: /dev/sda             # kernel name — rejected (not swap-stable)
         - name: /dev/disk/by-path/... # by-path — accepted (same-slot only)
   ```

   Should this be configurable, more permissive (user takes responsibility for any name), or stricter (reject `by-path` too)?

> 7. **Cross-host replacement for non-shared-metadata OSDs.** Same-host is required by this design because the captured `metadataVG` lives on the original host. For OSDs without a metadata device this argument doesn't apply. Ceph itself permits cross-host replacement: `ceph osd destroy` retains no host info; CRUSH auto-relocates the OSD on daemon start at the cost of full PG remapping. Should this flow be supported by Rook osd replacement?

Reviewer 6: Moving an OSD, which does not use split metadata, to a different host should be working now? What additional support is desired for doing that migration? Preserving the OSD ID?  Hmm. I wonder if migration is broken with encrypted OSD?

> 8. **Combine Capture and Replace into one Job.** The Capture step currently runs as a separate Job before the Replace Job. It could be merged into the Replace Job's first commands, with intermediate state persisted to a per-replacement ConfigMap (mirroring the prepare-job's CM hand-off pattern) so the Job remains retry-idempotent: on retry the Job reads the CM if present and skips re-capture, then proceeds with destroy + prepare. Saves one Job spawn per replacement; adds a CM-check branch at the top of the Job's bash and a CM-write helper invocation.

Reviewer 6: There is probably a larger question of what should Rook's ultimate state tracking of devices look like.  I confess that I have never used `ROOK_ENABLE_DISCOVERY_DAEMON` but I think one could make the argument that Rook should be actively tracking storage devices to make it easier to build features like this on top of state.  If Rook kept global state about devices it could potentially do things like automatically migrating OSDs between nodes. Or maybe device tracking should be a 3rd party controller that Rook builds on top of.

> 8. **Combine Capture and Replace into one Job.** The Capture step currently runs as a separate Job before the Replace Job. It could be merged into the Replace Job's first commands, with intermediate state persisted to a per-replacement ConfigMap (mirroring the prepare-job's CM hand-off pattern) so the Job remains retry-idempotent: on retry the Job reads the CM if present and skips re-capture, then proceeds with destroy + prepare. Saves one Job spawn per replacement; adds a CM-check branch at the top of the Job's bash and a CM-write helper invocation.

Reviewer 5: `capture` job is only `ceph-volume lvm list`.  So +1 on combining `capture` and `replace`.

## Proposed flow

This flow drives [Ceph's documented OSD-replacement procedure](https://docs.ceph.com/en/latest/rados/operations/add-or-rm-osds/#replacing-an-osd) (`ok-to-stop` → `osd out` → `safe-to-destroy` → `osd destroy` → `lvm zap` → `lvm batch --osd-ids`) from inside the existing CephCluster controller. The design adds one new Kubernetes Job — the Replace Job — that destroys the OSD in Ceph and zaps its data on the host.

State is read from Ceph and Kubernetes, not stored in Rook. Three markers carry it through the procedure:

- **Annotation on the OSD Deployment.** Marks intent. Stays in place while the operator validates the request and runs `ok-to-stop`, `osd out`, and `safe-to-destroy`. Cleaned up automatically when the Deployment is deleted.
- **[Suspended](https://kubernetes.io/blog/2021/04/12/introducing-suspended-jobs/#api-changes) Replace Job.** The operator captures the OSD's fsid, CV mode, data-device path, and encryption flag into the Job's env, creates the Job suspended, and deletes the OSD Deployment. The Job is resumed once the OSD pod terminates.

> - **Destroyed slot in the OSDMap.** The Replace Job runs `ceph osd destroy` to mark the OSD's slot as destroyed in the OSDMap and exits after cleanup is done. The user can physically swap the failed disk at any time after that — destroy and provision are decoupled. The existing prepare-job is extended to check the OSDMap for destroyed OSDs on the same node and reuse them when provisioning a new disk, the same primitive [cephadm uses for `orch osd rm --replace`](https://docs.ceph.com/en/latest/cephadm/services/osd/#replacing-an-osd).

Reviewer 2: Before the disk is physically replaced, how will Rook know to avoid re-creating an OSD on the disk that no longer has an OSD?

Author: right. this is a big gap in my design. in ceph-adm user initates disk destroy and provision processes manually. there is no autodetection of disk replacement. 
to avoid re-creating old OSD in rook we will have to store disk udev data to ignore it in prepare-job **OR** skip prepare-job  until we receive user confirmation that disk was replaced. At this point old osd deployment is already deleted so we cannot use it for annotations. Maybe user can annotate replace job or Cluster CR.

Author: @travisn @BlaineEXE i checked "dont't zap data lv/device to use it as a marker for disk replacement detection" approach and it looks promising.

## changes to proposed Replace Job:

The only thing we remove from the Job is the zap of the failed **data** device - we leave its ceph signature in place as the "disk still in the chassis" marker. Everything else stays: `ceph osd destroy` runs and marks osd as destroyed in ceph OSDMap; for shared-metadata the **DB LV** on the surviving metadata disk is still zapped (without `--destroy`, so it can be reused); for encrypted OSDs the data dm-crypt mapping is still closed.

## Prepare job

**Provisioning onto the failed disk is already suppressed - no change needed.** The device gate [`getAvailableDevices`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/daemon/ceph/osd/daemon.go#L361-L391) already skips any device that still carries a signature (filesystem / LVM / BlueStore label). So the failed disk is never picked for a new OSD while its marker is present:

| OSD type (host-based) | signature left on failed disk | skipped by the gate via |
|---|---|---|
| lvm, single, plain | data LV (LVM PV) | `FSTYPE=LVM2_member` |
| lvm, single, encrypted * | data LV (LVM PV) + LUKS | `FSTYPE=LVM2_member` |
| lvm, shared metadata, plain | data LV (LVM PV) | `FSTYPE=LVM2_member` |
| lvm, shared metadata, encrypted | data LV (LVM PV) + LUKS | `FSTYPE=LVM2_member` |
| raw, single, plain | BlueStore device label | `FSTYPE=ceph_bluestore` + bluestore-label check |

\* encryption forces lvm mode so "raw + encrypted"=="lvm, single, encrypted"

**One change is required.** Separately, the prepare job recreates Deployments for OSDs it finds via `ceph-volume lvm/raw list`, filtered only by cluster FSID. We add an analogous skip for destroyed slots: 
- for lvm [`volume.go#L1100-L1111`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/daemon/ceph/osd/volume.go#L1100-L1111) 
- for raw [`volume.go#L1300`](https://github.com/rook/rook/blob/59ce48ae88e5ea59df44249b41a887af96a2806c/pkg/daemon/ceph/osd/volume.go#L1300-L1310)

```diff
 	for name, osdInfo := range cephVolumeResult {
 		id, err := strconv.Atoi(name)
 		if err != nil {
 			logger.Errorf("bad osd returned from ceph-volume %q", name)
 			continue
 		}
+		// destroyedOSDs = ids from `ceph osd tree --states destroyed`.
+		// A destroyed slot whose on-disk LV/label still exists is a replacement marker for a failed disk, not a live OSD.
+		// It gets reused once the disk is physically swapped.
+		if destroyedOSDs[id] {
+			logger.Infof("skipping osd.%d: destroyed in the OSDMap, awaiting disk replacement", id)
+			continue
+		}
 		var osdFSID, osdDeviceClass string
 		for _, osd := range osdInfo {
 			if osd.Tags.ClusterFSID != cephfsid {
 				logger.Infof("skipping osd%d running on a different ceph cluster", id)
 				continue
 			}
```

And we need prepare Job adjustments from current design doc to reuse --osd-id and destroyed slot from shared metadata device.

Reviewer 2: That's great to see that if the OSD is just not wiped, we can plan to skip its OSD creation or reconcile. This seems like it should work very well. Perhaps it even means that we don't require a separate annotation for the OSD. If the osd prepare job finds an OSD is destroyed, and there is a disk available, then it can provision the OSD in its place.

Reviewer 2: After the portion of the metadata device is wiped, how do we know that it is available for a new OSD? Does Rook need to keep track of that, or perhaps ceph-volume can detect an available slot for a new OSD?

Author: > After the portion of the metadata device is wiped, how do we know that it is available for a new OSD? Does Rook need to keep track of that, or perhaps ceph-volume can detect an available slot for a new OSD?

i think in this case it will be better to also not zap db lv. it will be better to find it by destroyed osd-id from Ceph OSDMap, locate old db volume by osd id with `lvm list` and reuse. Otherwise we will have to calculate dbSize from sibling volumes because it is optional in spec.

Author: @travisn i updated doc accordingly but now i see one UX problem about how we signal that disk can be swapped to the user.

User allowed to swap the disk only after OSD is fully destroyed. Destroy commands like `osd destroy` and  closing dm-crypt mappings for encrypted OSD have to be called only after OSD daemon is down (deployment deleted). Only then device can be safely swapped.

So in current design proposal user cannot swap disk right away and cannot swap after OSD deployment was deleted. Instead user must lookup k8s job by name and check its status.

I think that it is not very convenient contract for human operator or for potential automation.

We also dont want to use CR status for indication which will keep replacement history forever. K8s Events also not reliable because they automatically deleted after some time. CM is a bit better because we can delete it but i think that it is also not the best contract to commuicate replacement status to the user.

I see 2 leternative options to communicate `ok-to-swap` to the user:
1. **Absence of deployment:** set `replicas=0`, wait pod destroyed, execute destroy commands, and only then delete the deployment.
2. **Annotation "ok-to-swap" on downscaled OSD deployment:** set `replicas=0`, wait pod destroyed, execute destroy commands, and then add `ok-to-swap` annotaion to OSD deployment. Downscaled deployment kept until new disk is provisioned, then updated to new spec with `replicas=1` and no annotaions.

the problem is that both approaches introduce a new valid sate `replicas=0` for OSD deployment.
in the first case downscaled deployment is relatively short-lived but we should not rely on timings.

It will not affect PDB because we marked osd out and destroyed only after data was redistributed(ok-to-destroy), but we will have to update health monitor to be aware of `replicas=0` OSD deployments (now it checks deployment `ReadyReplicas`). I dont think that changes to current rook code will be big. I think the bigger problem is having a new a new valid state `replicas=0 ` for OSD deployment will increase project complexity. I dont know if it is ok from maintainers perspective.

Reviewer 2: To signal that the disk is "ok-to-swap" can have several indicators. When the phase 1 job is completed, there will be several side effects that seem sufficient to indicate to the admin. 
1. The phase 1 job is in completed status. If desired, the user can query for job completion based on the OSD ID label. 
2. The OSD deployment is scaled down, has the `do-not-reconcile` label, and also a label or annotation that the device is safe to replace.

Since the `do-not-reconcile` label would be on the OSD, this is a valid state Rook handles today, so I am not concerned with maintaining this approach.

PDBs are not be impacted by an OSD that is scaled down. The PDB is only looking at OSD pods, which don't exist anymore for the destroyed OSD.

Author: ok, i personally think that keeping downscaled osd deployment with `do-not-reconcile` label and annotate it when it is  "ok-to-swap" is a better api rather than checking for phase 1 job.

Author: updated the doc  to use annotaion to signal ok-to-swap back to the user.

### Sequence

```mermaid

> sequenceDiagram

Reviewer 2: In this sequence, I understand there are two main phases of the replacement:
1. Mark the OSD for replacing (including waiting for safe-to-destroy, then destroy it)
2. After the disk is physically replaced, during the next reconcile the osd prepare job will re-create the OSD with the same ID. 

I would propose that the first step does not need to be part of the rook operator at all. This can be implemented as a job, similar to one of these:
- [osd-purge.yaml](https://github.com/rook/rook/blob/master/deploy/examples/osd-purge.yaml) job: Starts a job to run a special entry point in the rook operator image to purge an OSD. Flexibility to have all the custom code needed.
- [toolbox-job.yaml](https://github.com/rook/rook/blob/master/deploy/examples/toolbox-job.yaml): Runs a job with a shell script that can run arbitrary ceph commands. This may be simplest, to implement marking the OSD as out, wait for it to be safe-to-destroy, destroy it, etc.

After that job is completed, then the admin would know they can physically replace the devices to proceed to the second phase.

Reviewer 2: We might consider applying the `ceph.rook.io/do-not-reconcile` annotation to the OSD during the first phase, to avoid Rook from trying to reconcile it while it is being destroyed.

Reviewer 6: What is the motivation for trying to preserve the osd id?

Author: > What is the motivation for trying to preserve the osd id?

If we know that OSD is going to be replaced it is better to keep its ID in CRUSH map. Otherwise, purging and adding new OSD would change the weights of some CRUSH buckets (host, rack, etc.) and trigger unnecessary data movement.

Author: > I would propose that the first step does not need to be part of the rook operator at all. This can be implemented as a job, similar to one of these

Yes, it can be done. We will have to drop cluster CR validation or configure RBAC for job. But what are benefits of not having it in the operator? Operator reads annotation and spawns the job for user and it is easier for operator to do extra validation, enforce serial replacement order and do coordination with prepare job. 

You discovered a real gap in the design that after job finished, disk without OSD will look like a new disk for rook and prepare job  will try to create a new OSD from it. So now we will need some coordination between jobs to make replacement process work.

Maybe instead of deleting old OSD deployment operator can scale it to zero and skip prepare job on the node until user will swap the disk and add another annotation to the deployment?

Reviewer 2: That's fair, I can see the value in having the full solution in the operator. Perhaps we can just still think of it in two phases, where the new job will perform phase 1 with the OSD destroy and wait, and the OSD prepare job will complete phase 2 when the new physical disk is available.

4. **Device references resolve now.** Best-effort check of the OSD's `spec.storage` references, per [Device references may not survive a swap](#device-references-may-not-survive-a-swap). A reference that does not resolve is reported; the flow still proceeds.
5. **Drain admission.** `ceph osd ok-to-stop <id>` must return OK. Failure means removing this OSD would push some PG below `min_size`; the user resolves the underlying cluster state before re-annotating.

On all checks passing, the controller creates the Replace Job and returns.

#### 3. Replace Job: drain and destroy

> The operator creates a long-running Replace Job and returns. The Job reuses the prepare-job pod scaffold with a separate entry point, runs on the OSD's node, and carries `ttlSecondsAfterFinished` so it is removed after it completes. Its ServiceAccount is permitted to patch the OSD Deployment. All steps are idempotent on pod restart.

Reviewer 2: The prepare jobs by design are left in completed state, in case the logs are needed for troubleshooting. We might consider the same pattern for the replace job without `ttlSecondsAfterFinished`, in case troubleshooting is needed.

Author: agree. should operator delete it if there are no annotated deployment anymore?

Reviewer 2: The pattern for the osd prepare job is that it remains until the operator decides the osd prepare job needs to be run again (e.g. at the next reconcile). In the case of the replace job, I think this pattern works well. We could leave the job until the time when more OSDs need to be replaced, then we run a new job.

The prepare-job writes OSDInfo to the existing per-node status CM `rook-ceph-osd-<node>-status`. The operator then clears the replacement annotations and the `do-not-reconcile` label, which returns the OSD to the normal update path: the Deployment scales back to `replicas=1` with the new disk's OSDInfo. The controller polls `ceph osd tree` each reconcile until `osd.<id>` is `up` AND `in`; once observed, it emits an `OSDReplaceCompleted` event on the deployment.

**Triggering the reprovision after a swap.** The prepare-job that detects the new disk runs only during a CephCluster reconcile. Rook has no active swap detection, and a successful reconcile schedules no periodic resync, so a disk swapped while the operator is idle is not noticed until a reconcile is triggered — by the `rook-discover` daemon when it is enabled, or by a manual operator restart or CephCluster spec edit. See [open question 3](#open-questions).

### Coordination

> Replacements run serially per CephCluster, matching cephadm's `osd rm` queue and Rook's OSD migration. On each reconcile the operator looks for an in-progress replacement — an OSD Deployment carrying the `osd.rook.io/replace` annotation whose OSD is not yet back `up` and `in` — and creates at most one Replace Job at a time, deferring any other annotated OSDs to later reconciles. The in-progress replacement is identified from its own markers (the annotated Deployment, the Replace Job, and the destroyed slot), so unlike migration — which records the in-flight OSD id in a ConfigMap — no separate serialization state is needed.

Reviewer 2: Why not run start the replace job for all annotated OSDs in parallel? The risk seems low of some parallel issue. If the user annotates too many OSDs, they might just never be safe to destroy.

Author: good point. i checked how it is done in ceph-adm. it has `max_osd_draining_count` config (10 by default), so  its remove osd queue is not serial. But it is not clear to me how we can implement batching inside the job. Job is per-node. In this case it makes sense to keep draining logic in operator:
- list all annotated osd deployments
- put all osd id batch into `osd ok-to-stop` and `safe-to-destroy`
- then spawn per osd jobs on nodes to run cleanup logic on hosts.

Reviewer 2: The job is per-node only for the encrypted osd teardown? I wonder if that should be part of the wipe process that is outside of Rook's scope? Or perhaps start a per-node job after they are destroyed, only for encrypted OSDs. But I'm not clear of the implications there. Perhaps handling encrypted OSDs could even come in a follow-up PR if that helps this main design move forward sooner.

Otherwise, seems like we can keep the job and give it a list of OSDs. The job could still mark them as out, wait for them all to be safe-to-destroy (with a single command that is queried regularly), and then finally destroy them.

Author: but in case if we pass osdIDs to job they cannot be changed. so if user will annotate osds one by one, operator will create job for every osd and they will call ok-to-stop and safe-to-destroy with a single id in parallel. Also jobs are per-node and for parallel destroy we need to pass all ids to ceph commands.

on the other hand, if we keep draining in operator, then we can list all annotated osds on every reconcile and use it in safe-to-destroy and then create a job.

Reviewer 2: All of these questions made me realize that there must be a simpler architecture that avoids many of these issues. Here is another approach to consider...

Rook has an [OSD health goroutine](https://github.com/rook/rook/blob/8cbec942f48778bc04d51a90abf8d863339d4ac2/pkg/operator/ceph/cluster/osd/health.go#L72) that periodically checks the health of the OSDs. The routine even already checks the [osd dump](https://github.com/rook/rook/blob/8cbec942f48778bc04d51a90abf8d863339d4ac2/pkg/operator/ceph/cluster/osd/health.go#L146-L184), which tells status about the OSDs being `up`, or `out`. 

This goroutine seems a perfect fit for everything the replace job was going to do. That goroutine can do the following:
1. If an OSD has the replace annotation:
   - Add the `do-not-reconcile` label so the main reconcile will ignore the OSD
   - Mark the osd `out`
2. If all of the OSDs marked for replacement are `safe-to-destroy`:
   - Destroy each OSD
   - Scale down the OSD deployment
   - If cleanup is needed on the host, start a job on that node (e.g. encrypted OSDs clean up the host)
3. If not `safe-to-destroy`, wait 60s and check again

The advantages of this approach:
- Normal reconcile does not perform anything related to OSD replacement
- There is no separate job needed for replacement, unless needed for node cleanup
- As additional OSDs are marked for replacement, they will be immediately be picked up by the next health check (every 60s)

Reviewer 2: One more thought... If an OSD is marked `out`, it seems like it should be safe to query `safe-to-destroy` individually for an OSD, as long as we query `safe-to-destroy` again between destroying one OSD and deciding if the next OSD can be destroyed. Since `out`, they would not ever be assigned more data or PGs. 

If multiple OSDs are marked for replace, it's possible some OSDs could be safe to destroy, while others are not. All of them don't need to wait to be destroyed. 

But if I'm missing something about safe-to-destroy, please correct me.

Author: agree. moved destroy logic to osd health goroutine

1. **Metadata device: reuse the surviving DB LV, or consult the spec?** The design recovers the metadata device from the destroyed OSD's surviving DB LV (`ceph-volume lvm list <id>`) and reuses that LV in place. This is drift-proof and needs no spec read. The trade-off: a user edit to `spec.storage[<node>].config.metadataDevice` is ignored for the replacement. (A replacement where no DB LV survives is already a non-goal.) Should provisioning ever consult the spec, or always follow the surviving DB LV?

2. **Where to report the validation result.** The current proposal emits a Warning event on the CephCluster CR (matching Rook convention). Should it also emit on the OSD Deployment, the object the caller annotated?

> 3. **Should Rook add a trigger for the reprovision after the swap?** The swap is detected only on a CephCluster reconcile, which Rook does not trigger on its own. With the discovery daemon enabled this happens automatically; without it, the user must restart the operator, so replacement is not fully automated. Directions:

Reviewer 2: I would vote to document the usage of the discovery daemon if they want to automate the replacing after physical swap, or else they can just restart the operator.

Author: ok
